File: lib/changelog/src/propertyChanges.ts
import { createHash } from 'node:crypto';
import { toSlug } from '@deadlog/utils';

export { PROPERTY_EXTRACTION_VERSION } from '@deadlog/utils';

export const LINKED_PROPERTIES = ['cooldown', 'radius', 'damage'] as const;
export type LinkedProperty = (typeof LINKED_PROPERTIES)[number];

export interface PropertyValue {
	amount: number;
	unit: string;
	text: string;
}

export interface PropertyEvent {
	property: LinkedProperty;
	qualifier: string;
	old: PropertyValue;
	new: PropertyValue;
}

export type BulletReading =
	| { kind: 'event'; event: PropertyEvent }
	| { kind: 'barrier'; properties: LinkedProperty[] | 'all' }
	| { kind: 'unrelated' };

const VERBS =
	'(?:increased|reduced|decreased|lowered|raised|changed|adjusted|improved|nerfed|buffed)';
const VALUE = '([+-]?\\d+(?:\\.\\d+)?)(s|m|%)?';
const EVENT_RE = new RegExp(
	`^(?:(T[123]) )?(base )?(cooldown|radius|damage) (?:has been |was |is )?${VERBS} from ${VALUE} to ${VALUE}$`,
	'i'
);
const REWORK_RE =
	/\b(?:rework(?:ed)?|redesigned|replaced|renamed|removed|new ability|now a (?:passive|active)|no longer)\b/i;

function value(amount: string, unit: string | undefined, other: string | undefined) {
	return {
		amount: Number(amount),
		unit: unit ?? other ?? '',
		text: `${amount}${unit ?? ''}`
	};
}

export function readBullet(bullet: string, scopeName: string | null): BulletReading {
	let text = bullet.trim().replace(/\.$/, '');
	if (scopeName) {
		if (!text.toLowerCase().startsWith(`${scopeName.toLowerCase()} `)) {
			return mentions(text);
		}
		text = text.slice(scopeName.length + 1);
	}

	const match = EVENT_RE.exec(text);
	if (!match) return mentions(text);

	const [, tier, base, property, oldAmount, oldUnit, newAmount, newUnit] = match;
	if (oldUnit && newUnit && oldUnit !== newUnit) return mentions(text);
	return {
		kind: 'event',
		event: {
			property: property.toLowerCase() as LinkedProperty,
			qualifier: [tier?.toUpperCase(), base ? 'base' : null].filter(Boolean).join(' '),
			old: value(oldAmount, oldUnit, newUnit),
			new: value(newAmount, newUnit, oldUnit)
		}
	};
}

function mentions(text: string): BulletReading {
	if (REWORK_RE.test(text)) return { kind: 'barrier', properties: 'all' };
	const properties = LINKED_PROPERTIES.filter((property) =>
		new RegExp(`\\b${property}\\b`, 'i').test(text)
	);
	return properties.length > 0 ? { kind: 'barrier', properties } : { kind: 'unrelated' };
}

export interface ScopedBullet {
	patchId: string;
	publishedAt: string;
	entityType: 'hero' | 'item';
	entityId: number;
	ability: string | null;
	abilitySlug: string | null;
	groupIndex: number;
	bulletIndex: number;
	text: string;
}

export interface LinkedEvent extends PropertyEvent {
	patchId: string;
	entityType: 'hero' | 'item';
	entityId: number;
	abilitySlug: string | null;
	groupIndex: number;
	bulletIndex: number;
	digest: string;
	previous: { patchId: string; old: string; new: string } | null;
	unlinked: 'first' | 'barrier' | 'ambiguous-order' | 'value-mismatch' | null;
}

const digest = (text: string): string =>
	createHash('sha256').update(text).digest('hex').slice(0, 16);

interface Scope {
	key: string;
	name: string | null;
	trusted: boolean;
}

function scopeOf(bullet: ScopedBullet): Scope | null {
	if (bullet.entityType === 'item') {
		return {
			key: `item:${bullet.entityId}`,
			name: bullet.ability,
			trusted: !bullet.ability
		};
	}
	if (!bullet.ability || !bullet.abilitySlug) return null;
	return {
		key: `hero:${bullet.entityId}:${bullet.abilitySlug}`,
		name: bullet.ability,
		trusted: toSlug(bullet.ability) === bullet.abilitySlug
	};
}

export function linkPropertyChanges(bullets: ScopedBullet[]): LinkedEvent[] {
	const ordered = [...bullets].sort(
		(a, b) =>
			a.publishedAt.localeCompare(b.publishedAt) ||
			a.patchId.localeCompare(b.patchId) ||
			a.groupIndex - b.groupIndex ||
			a.bulletIndex - b.bulletIndex
	);
	const tiedTimes = new Set<string>();
	const seenTimes = new Map<string, string>();
	for (const bullet of ordered) {
		const owner = seenTimes.get(bullet.publishedAt);
		if (owner && owner !== bullet.patchId) tiedTimes.add(bullet.publishedAt);
		seenTimes.set(bullet.publishedAt, bullet.patchId);
	}

	interface Step {
		patchId: string;
		tied: boolean;
		events: LinkedEvent[];
		barrier: boolean;
	}
	const chains = new Map<string, Step[]>();
	const stepFor = (chain: string, bullet: ScopedBullet): Step => {
		const steps = chains.get(chain) ?? [];
		chains.set(chain, steps);
		let step = steps.at(-1);
		if (!step || step.patchId !== bullet.patchId) {
			step = {
				patchId: bullet.patchId,
				tied: tiedTimes.has(bullet.publishedAt),
				events: [],
				barrier: false
			};
			steps.push(step);
		}
		return step;
	};

	const events: LinkedEvent[] = [];
	const chainsByScope = new Map<string, Set<string>>();
	for (const bullet of ordered) {
		const scope = scopeOf(bullet);
		if (!scope) continue;
		const read = readBullet(bullet.text, scope.name);
		const reading: BulletReading =
			read.kind === 'event' && !scope.trusted
				? { kind: 'barrier', properties: [read.event.property] }
				: read;
		if (reading.kind === 'unrelated') continue;

		const known = chainsByScope.get(scope.key) ?? new Set<string>();
		chainsByScope.set(scope.key, known);

		if (reading.kind === 'barrier') {
			for (const chain of known) {
				const [, property] = chain.split('|');
				if (
					reading.properties === 'all' ||
					reading.properties.includes(property as LinkedProperty)
				) {
					stepFor(chain, bullet).barrier = true;
				}
			}
			continue;
		}

		const chain = `${scope.key}|${reading.event.property}|${reading.event.qualifier}`;
		known.add(chain);
		const event: LinkedEvent = {
			...reading.event,
			patchId: bullet.patchId,
			entityType: bullet.entityType,
			entityId: bullet.entityId,
			abilitySlug: bullet.abilitySlug,
			groupIndex: bullet.groupIndex,
			bulletIndex: bullet.bulletIndex,
			digest: digest(bullet.text),
			previous: null,
			unlinked: 'first'
		};
		stepFor(chain, bullet).events.push(event);
		events.push(event);
	}

	for (const steps of chains.values()) {
		let last: Step | null = null;
		let blocked = false;
		for (const step of steps) {
			const ordered = !step.tied && step.events.length === 1;
			for (const event of step.events) {
				if (!last) event.unlinked = 'first';
				else if (blocked || step.barrier) event.unlinked = 'barrier';
				else if (!ordered || last.tied || last.events.length !== 1) {
					event.unlinked = 'ambiguous-order';
				} else {
					const [before] = last.events;
					const sameUnit =
						event.old.unit === before.new.unit || !event.old.unit || !before.new.unit;
					if (event.old.amount !== before.new.amount || !sameUnit) {
						event.unlinked = 'value-mismatch';
					} else {
						event.unlinked = null;
						event.previous = {
							patchId: before.patchId,
							old: before.old.text,
							new: before.new.text
						};
					}
				}
			}
			if (step.events.length > 0) {
				last = step;
				blocked = step.barrier;
			} else if (step.barrier) blocked = true;
		}
	}
	return events;
}

File: crates/changelog/src/property_changes.rs
use std::collections::{HashMap, HashSet};
use std::sync::LazyLock;

use deadlog_model::{EntityType, is_js_space, to_slug};
use regex::Regex;
use sha2::{Digest, Sha256};

/// Shared event contract: readers must reject links from an unknown extractor.
pub const PROPERTY_EXTRACTION_VERSION: i64 = 1;

pub const LINKED_PROPERTIES: [&str; 3] = ["cooldown", "radius", "damage"];

#[derive(Debug, Clone, PartialEq)]
pub struct PropertyValue {
    pub amount: f64,
    pub unit: String,
    pub text: String,
}

#[derive(Debug, Clone, PartialEq)]
pub struct PropertyEvent {
    pub property: &'static str,
    pub qualifier: String,
    pub old: PropertyValue,
    pub new: PropertyValue,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Barrier {
    All,
    Properties(Vec<&'static str>),
}

#[derive(Debug, Clone, PartialEq)]
pub enum BulletReading {
    Event(PropertyEvent),
    Barrier(Barrier),
    Unrelated,
}

const VERBS: &str = "(?:increased|reduced|decreased|lowered|raised|changed|adjusted|improved|nerfed|buffed)";
const VALUE: &str = r"([+-]?[0-9]+(?:\.[0-9]+)?)(s|m|%)?";

static EVENT_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(&format!(
        r"(?i)^(?:(T[123]) )?(base )?(cooldown|radius|damage) (?:has been |was |is )?{VERBS} from {VALUE} to {VALUE}$"
    ))
    .unwrap()
});
static REWORK_RE: LazyLock<Regex> = LazyLock::new(|| {
    Regex::new(
        r"(?i)(?-u:\b)(?:rework(?:ed)?|redesigned|replaced|renamed|removed|new ability|now a (?:passive|active)|no longer)(?-u:\b)",
    )
    .unwrap()
});
static PROPERTY_RES: LazyLock<Vec<(&'static str, Regex)>> = LazyLock::new(|| {
    LINKED_PROPERTIES
        .iter()
        .map(|property| (*property, Regex::new(&format!(r"(?i)(?-u:\b){property}(?-u:\b)")).unwrap()))
        .collect()
});

fn value(amount: &str, unit: Option<&str>, other: Option<&str>) -> PropertyValue {
    PropertyValue {
        amount: amount.parse().unwrap_or(f64::NAN),
        unit: unit.or(other).unwrap_or_default().to_string(),
        text: format!("{amount}{}", unit.unwrap_or_default()),
    }
}

/// `text.slice(units)` for a UTF-16 offset.
fn slice_utf16(text: &str, units: usize) -> &str {
    let mut count = 0;
    for (index, c) in text.char_indices() {
        if count >= units {
            return &text[index..];
        }
        count += c.len_utf16();
    }
    ""
}

pub fn read_bullet(bullet: &str, scope_name: Option<&str>) -> BulletReading {
    let trimmed = bullet.trim_matches(is_js_space);
    let mut text = trimmed.strip_suffix('.').unwrap_or(trimmed);
    if let Some(scope) = scope_name.filter(|scope| !scope.is_empty()) {
        if !text.to_lowercase().starts_with(&format!("{} ", scope.to_lowercase())) {
            return mentions(text);
        }
        text = slice_utf16(text, scope.encode_utf16().count() + 1);
    }

    let Some(captures) = EVENT_RE.captures(text) else { return mentions(text) };
    let group = |index| captures.get(index).map(|m| m.as_str());
    let (old_unit, new_unit) = (group(5), group(7));
    if let (Some(old), Some(new)) = (old_unit, new_unit)
        && old != new
    {
        return mentions(text);
    }
    let property = group(3).unwrap_or_default().to_lowercase();
    let property = LINKED_PROPERTIES.into_iter().find(|linked| *linked == property).unwrap_or("cooldown");
    let qualifier: Vec<String> =
        [group(1).map(str::to_uppercase), group(2).map(|_| "base".to_string())].into_iter().flatten().collect();
    BulletReading::Event(PropertyEvent {
        property,
        qualifier: qualifier.join(" "),
        old: value(group(4).unwrap_or_default(), old_unit, new_unit),
        new: value(group(6).unwrap_or_default(), new_unit, old_unit),
    })
}

fn mentions(text: &str) -> BulletReading {
    if REWORK_RE.is_match(text) {
        return BulletReading::Barrier(Barrier::All);
    }
    let properties: Vec<&'static str> =
        PROPERTY_RES.iter().filter(|(_, pattern)| pattern.is_match(text)).map(|(property, _)| *property).collect();
    if properties.is_empty() { BulletReading::Unrelated } else { BulletReading::Barrier(Barrier::Properties(properties)) }
}

#[derive(Debug, Clone, PartialEq)]
pub struct ScopedBullet {
    pub patch_id: String,
    pub published_at: String,
    pub entity_type: EntityType,
    pub entity_id: i64,
    pub ability: Option<String>,
    pub ability_slug: Option<String>,
    pub group_index: usize,
    pub bullet_index: usize,
    pub text: String,
}

#[derive(Debug, Clone, PartialEq)]
pub struct PreviousChange {
    pub patch_id: String,
    pub old: String,
    pub new: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Unlinked {
    First,
    Barrier,
    AmbiguousOrder,
    ValueMismatch,
}

#[derive(Debug, Clone, PartialEq)]
pub struct LinkedEvent {
    pub event: PropertyEvent,
    pub patch_id: String,
    pub entity_type: EntityType,
    pub entity_id: i64,
    pub ability_slug: Option<String>,
    pub group_index: usize,
    pub bullet_index: usize,
    pub digest: String,
    pub previous: Option<PreviousChange>,
    pub unlinked: Option<Unlinked>,
}

fn digest(text: &str) -> String {
    let hash = Sha256::digest(text.as_bytes());
    hash.iter().take(8).map(|byte| format!("{byte:02x}")).collect()
}

struct Scope {
    key: String,
    name: Option<String>,
    trusted: bool,
}

fn scope_of(bullet: &ScopedBullet) -> Option<Scope> {
    match bullet.entity_type {
        EntityType::Item => Some(Scope {
            key: format!("item:{}", bullet.entity_id),
            name: bullet.ability.clone(),
            trusted: bullet.ability.as_deref().is_none_or(str::is_empty),
        }),
        EntityType::Hero => {
            let ability = bullet.ability.as_deref().filter(|ability| !ability.is_empty())?;
            let slug = bullet.ability_slug.as_deref().filter(|slug| !slug.is_empty())?;
            Some(Scope {
                key: format!("hero:{}:{slug}", bullet.entity_id),
                name: Some(ability.to_string()),
                trusted: to_slug(ability) == slug,
            })
        }
    }
}

struct Step {
    patch_id: String,
    tied: bool,
    events: Vec<usize>,
    barrier: bool,
}

/// Links each readable property change to the one it continues, refusing whenever the
/// order or the values leave any doubt.
pub fn link_property_changes(bullets: &[ScopedBullet]) -> Vec<LinkedEvent> {
    let mut ordered: Vec<&ScopedBullet> = bullets.iter().collect();
    // localeCompare in the TypeScript; these ids and ISO stamps sort the same by bytes.
    ordered.sort_by(|a, b| {
        a.published_at
            .cmp(&b.published_at)
            .then_with(|| a.patch_id.cmp(&b.patch_id))
            .then(a.group_index.cmp(&b.group_index))
            .then(a.bullet_index.cmp(&b.bullet_index))
    });
    let mut tied_times = HashSet::new();
    let mut seen_times: HashMap<&str, &str> = HashMap::new();
    for bullet in &ordered {
        if let Some(owner) = seen_times.get(bullet.published_at.as_str())
            && *owner != bullet.patch_id
        {
            tied_times.insert(bullet.published_at.as_str());
        }
        seen_times.insert(&bullet.published_at, &bullet.patch_id);
    }

    let mut chain_order: Vec<String> = Vec::new();
    let mut chains: HashMap<String, Vec<Step>> = HashMap::new();
    let mut step_for = |chains: &mut HashMap<String, Vec<Step>>, chain: &str, bullet: &ScopedBullet| -> (String, usize) {
        let steps = chains.entry(chain.to_string()).or_insert_with(|| {
            chain_order.push(chain.to_string());
            Vec::new()
        });
        if steps.last().is_none_or(|step| step.patch_id != bullet.patch_id) {
            steps.push(Step {
                patch_id: bullet.patch_id.clone(),
                tied: tied_times.contains(bullet.published_at.as_str()),
                events: Vec::new(),
                barrier: false,
            });
        }
        (chain.to_string(), steps.len() - 1)
    };

    let mut events: Vec<LinkedEvent> = Vec::new();
    let mut chains_by_scope: HashMap<String, Vec<String>> = HashMap::new();
    for bullet in ordered {
        let Some(scope) = scope_of(bullet) else { continue };
        let reading = match read_bullet(&bullet.text, scope.name.as_deref()) {
            BulletReading::Event(event) if !scope.trusted => {
                BulletReading::Barrier(Barrier::Properties(vec![event.property]))
            }
            reading => reading,
        };
        let known = chains_by_scope.entry(scope.key.clone()).or_default();
        match reading {
            BulletReading::Unrelated => {}
            BulletReading::Barrier(barrier) => {
                for chain in known.iter() {
                    let property = chain.split('|').nth(1).unwrap_or_default();
                    let blocks = match &barrier {
                        Barrier::All => true,
                        Barrier::Properties(properties) => properties.contains(&property),
                    };
                    if blocks {
                        let (key, index) = step_for(&mut chains, chain, bullet);
                        chains.get_mut(&key).expect("chain exists")[index].barrier = true;
                    }
                }
            }
            BulletReading::Event(event) => {
                let chain = format!("{}|{}|{}", scope.key, event.property, event.qualifier);
                if !known.contains(&chain) {
                    known.push(chain.clone());
                }
                let (key, index) = step_for(&mut chains, &chain, bullet);
                chains.get_mut(&key).expect("chain exists")[index].events.push(events.len());
                events.push(LinkedEvent {
                    digest: digest(&bullet.text),
                    event,
                    patch_id: bullet.patch_id.clone(),
                    entity_type: bullet.entity_type,
                    entity_id: bullet.entity_id,
                    ability_slug: bullet.ability_slug.clone(),
                    group_index: bullet.group_index,
                    bullet_index: bullet.bullet_index,
                    previous: None,
                    unlinked: Some(Unlinked::First),
                });
            }
        }
    }

    for steps in chains.values() {
        let mut last: Option<&Step> = None;
        let mut blocked = false;
        for step in steps {
            let ordered = !step.tied && step.events.len() == 1;
            for &index in &step.events {
                let (unlinked, previous) = match last {
                    None => (Some(Unlinked::First), None),
                    Some(_) if blocked || step.barrier => (Some(Unlinked::Barrier), None),
                    Some(last) if !ordered || last.tied || last.events.len() != 1 => (Some(Unlinked::AmbiguousOrder), None),
                    Some(last) => {
                        let before = &events[last.events[0]];
                        let event = &events[index].event;
                        let same_unit = event.old.unit == before.event.new.unit
                            || event.old.unit.is_empty()
                            || before.event.new.unit.is_empty();
                        if event.old.amount != before.event.new.amount || !same_unit {
                            (Some(Unlinked::ValueMismatch), None)
                        } else {
                            let previous = PreviousChange {
                                patch_id: before.patch_id.clone(),
                                old: before.event.old.text.clone(),
                                new: before.event.new.text.clone(),
                            };
                            (None, Some(previous))
                        }
                    }
                };
                events[index].unlinked = unlinked;
                events[index].previous = previous;
            }
            if !step.events.is_empty() {
                last = Some(step);
                blocked = step.barrier;
            } else if step.barrier {
                blocked = true;
            }
        }
    }
    events
}
